### PR TITLE
Fix integer comparison warnings for equihash.cpp/h

### LINKED ISSUE #47: Tons of compile warnings in equihash.cpp (comparison between signed and unsigned integer expressions)
```
crypto/equihash.cpp: In instantiation of ‘bool Equihash<N, K>::BasicSolve(const eh_HashState&, std::function<bool(std::vector<unsigned char>)>, std::function<bool(EhSolverCancelCheck)>) [with unsigned int N = 96u; unsigned int K = 3u; eh_HashState = crypto_generichash_blake2b_state]’:
```

```
crypto/equihash.cpp: In instantiation of ‘bool Equihash<N, K>::IsValidSolution(const eh_HashState&, std::vector<unsigned char>) [with unsigned int N = 200u; unsigned int K = 9u; eh_HashState = crypto_generichash_blake2b_state]’:
crypto/equihash.cpp:785:111:   required from here
crypto/equihash.cpp:741:27: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
crypto/equihash.cpp: In instantiation of ‘bool Equihash<N, K>::IsValidSolution(const eh_HashState&, std::vector<unsigned char>) [with unsigned int N = 96u; unsigned int K = 5u; eh_HashState = crypto_generichash_blake2b_state]’:
crypto/equihash.cpp:795:110:   required from here
crypto/equihash.cpp:741:27: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
crypto/equihash.cpp: In instantiation of ‘bool Equihash<N, K>::IsValidSolution(const eh_HashState&, std::vector<unsigned char>) [with unsigned int N = 48u; unsigned int K = 5u; eh_HashState = crypto_generichash_blake2b_state]’:
crypto/equihash.cpp:805:110:   required from here
crypto/equihash.cpp:741:27: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
crypto/equihash.cpp: In instantiation of ‘FullStepRow<WIDTH>::FullStepRow(const FullStepRow<W>&, const FullStepRow<W>&, size_t, size_t, int) [with long unsigned int W = 22ul; long unsigned int WIDTH = 38ul; size_t = long unsigned int]’:
crypto/equihash.cpp:425:91:   required from ‘bool Equihash<N, K>::BasicSolve(const eh_HashState&, std::function<bool(std::vector<unsigned char>)>, std::function<bool(EhSolverCancelCheck)>) [with unsigned int N = 96u; unsigned int K = 3u; eh_HashState = crypto_generichash_blake2b_state]’
./crypto/equihash.h:220:67:   required from here
crypto/equihash.cpp:233:26: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
     for (int i = trim; i < len; i++)
                          ^
```

Never had this many compile warnings with Bitcoin Core or Bitcoin Cash.

### PR DESCRIPTION
**Pull request by wilsonmeier**
Integrated fixed from another PR and added more fixes
fixes #47

### CODE REVIEW

File: src/crypto/equihash.cpp [hunk 1]
```diff
@@ -230,12 +230,12 @@ FullStepRow<WIDTH>::FullStepRow(const unsigned char* hashIn, size_t hInLen,
 }
 
 template<size_t WIDTH> template<size_t W>
-FullStepRow<WIDTH>::FullStepRow(const FullStepRow<W>& a, const FullStepRow<W>& b, size_t len, size_t lenIndices, int trim) :
+FullStepRow<WIDTH>::FullStepRow(const FullStepRow<W>& a, const FullStepRow<W>& b, size_t len, size_t lenIndices, size_t trim) :
         StepRow<WIDTH> {a}
 {
     assert(len+lenIndices <= W);
     assert(len-trim+(2*lenIndices) <= WIDTH);
```
Comment (h4x3rotab) [inline src/crypto/equihash.cpp]: Just a reminder. I didn't check all the upstreams but changing `int` to `size_t` may introduce potential integer overflow when doing subtraction.

As an example: a simple `for (int i = 0; i < n; i++)` to `for (size_t i = 0; i < n; i++)` won't have problem but `for (size i = n-1; i >=0; i--)` is an error.

However in this case, it looks like everything is fine.

File: src/crypto/equihash.cpp [hunk 2]
```diff
@@ -511,7 +511,7 @@ bool Equihash<N,K>::OptimisedSolve(const eh_HashState& base_state,
 
     const eh_index soln_size { 1 << K };
     std::vector<std::shared_ptr<eh_trunc>> partialSolns;
-    int invalidCount = 0;
+    size_t invalidCount = 0;
```
Comment (h4x3rotab) [inline src/crypto/equihash.cpp]: Check L717 & L719. When printing either it should be `int invalidCount = 0;` or `printf("... %zu", invlidCount)`.

File: src/crypto/equihash.cpp [hunk 3]
```diff
@@ -531,29 +531,29 @@ bool Equihash<N,K>::OptimisedSolve(const eh_HashState& base_state,
         }
 
         // 3) Repeat step 2 until 2n/(k+1) bits remain
-        for (int r = 1; r < K && Xt.size() > 0; r++) {
+        for (size_t r = 1; r < K && Xt.size() > 0; r++) {
             LogPrint(BCLog::POW, "Round %d:\n", r);
```
Comment (h4x3rotab) [inline src/crypto/equihash.cpp]: `LogPrint(BCLog::POW, "Round %zu:\n", r);`

File: src/leveldb/util/logging.cc [hunk 4]
```diff
@@ -52,7 +52,7 @@ bool ConsumeDecimalNumber(Slice* in, uint64_t* val) {
     unsigned char c = (*in)[0];
     if (c >= '0' && c <= '9') {
       ++digits;
-      const int delta = (c - '0');
```
Comment (h4x3rotab) [inline src/leveldb/util/logging.cc]: I believe this file was copied from LevelDB. Let's keep it as is.

File: src/crypto/equihash.cpp [hunk 5]
```diff
@@ -230,12 +230,12 @@ FullStepRow<WIDTH>::FullStepRow(const unsigned char* hashIn, size_t hInLen,
...
 
 template<size_t WIDTH> template<size_t W>
-FullStepRow<WIDTH>::FullStepRow(const FullStepRow<W>& a, const FullStepRow<W>& b, size_t len, size_t lenIndices, int trim) :
+FullStepRow<WIDTH>::FullStepRow(const FullStepRow<W>& a, const FullStepRow<W>& b, size_t len, size_t lenIndices, size_t trim) :
         StepRow<WIDTH> {a}
 {
     assert(len+lenIndices <= W);
     assert(len-trim+(2*lenIndices) <= WIDTH);
-    for (int i = trim; i < len; i++)
+    for (size_t i = static_cast<size_t>(trim); i < len; i++)
```
Comment (h4x3rotab) [inline src/crypto/equihash.cpp]: `trim` is already `size_t`. So I think there's no need to cast again.

File: src/crypto/equihash.cpp [hunk 6]
```diff
@@ -352,28 +352,28 @@ bool Equihash<N,K>::BasicSolve(const eh_HashState& base_state,
     }
 
     // 3) Repeat step 2 until 2n/(k+1) bits remain
-    for (int r = 1; r < K && X.size() > 0; r++) {
+    for (unsigned int r = 1; r < K && X.size() > 0; r++) {
         LogPrint(BCLog::POW, "Round %d:\n", r);
```
Comment (h4x3rotab) [inline src/crypto/equihash.cpp]: `LogPrint(BCLog::POW, "Round %u:\n", r);`

File: src/crypto/equihash.cpp (hunk 5 above)
Comment (wilsonmeier) [reply src/crypto/equihash.cpp]: Correct, will remove the cast.

**Pull request merged**